7659 7661 TOGGLE Betinget visning av Trygdeavgift- og Vedtak-steg

ragnarwestad opened the pull request:
## Implementert dynamisk visning av Trygdeavgift- og Vedtak-steg

  ### 🎯 Hovedproblemer
  1. **Trygdeavgift-steget** skulle kun vises når det er relevant (når det skal beregnes trygdeavgift)
  2. **Vedtak-steget** dukket ikke opp umiddelbart når Trygdeavgift-steget ble godkjent
     - Det viste seg først ved navigering til et annet steg eller ved klikk på "Bekreft og fortsett"
  3. **Radio-knapp i Trygdeavgift** hoppet tilbake til "Ja" etter klikk på "Nei"
  4. **Doble API-kall** ved brukerinteraksjon i Trygdeavgift-steget

  ### 🔍 Rotårsak
  1. Ingen mekanisme for å styre når Trygdeavgift-steget skulle vises/skjules
  2. `oppdaterStegData()` i Stegvelger kalte kun `oppdaterAktuelleSteg()` når `oppdaterRedux=true`
  3. Trygdeavgift brukte `oppdaterRedux=false` for å unngå unødvendige re-renders som resetter komponentens state
  4. Når trygdeavgiftStatus endret seg, ble ikke steg-kjeden rebuildet
  5. **Validering i Trygdeavgift kjørte før data var lastet fra backend**, noe som førte til:
     - Form-reset som overskrev brukerens valg
     - Doble API-kall
     - Radio-knapp som hoppet tilbake

  ### ✅ Løsning

  #### Betinget visning av Trygdeavgift-steg
  - Flyttet `vurderingTrygdeavgift` fra `src/sider/ftrl/` til `src/felleskomponenter/trygdeavgift/` siden den brukes i både FTRL og EU/EØS
  - Periode-steget returnerer nå Trygdeavgift som `nesteSteg` kun når lovvalgsbestemmelse er satt
  - Trygdeavgift-steget vises dynamisk basert på om Periode er fullført

  #### Betinget visning av Vedtak-steg
  - Opprettet `trygdeavgiftStatus` store i Stegvelger
  - Trygdeavgift-komponenten kaller `oppdaterStatus(isValid)` basert på validering
  - Status lagres uten Redux-update (`oppdaterRedux: false`) for å unngå re-render
  - Vedtak-steget vises kun når Trygdeavgift er valid (`trygdeavgiftStatus === true`)

  #### Dynamisk steg-kjede rebuild
  - La til `else`-branch i `oppdaterStegData()` som rebuilder steg-kjeden
  - Kun for `type="trygdeavgiftStatus"` for å unngå concurrency-problemer
  - Bruker `debouncedOppdaterAktuelleSteg()` (300ms) for å unngå multiple API-kall
  - Steg-kjeden rebuildes nå når `trygdeavgiftStatus` endres

  #### Fikset radio-knapp og doble API-kall i Trygdeavgift
  - La til `dataErLastet` state for å tracke når initial data er hentet fra backend
  - Validering kjører kun når `dataErLastet=true`, slik at data er klar før validering
  - La til `skalResetteFormverdier` parameter for å unngå å overskrive brukerens endringer
  - Resulterer i:
    - ✅ Radio-knappen hopper ikke lenger tilbake
    - ✅ Kun ett API-kall ved interaksjon
    - ✅ Validering kjører kun når data er klar

  #### Type-sikkerhet i Trygdeavgift
  - Fjernet alle `any`-typer fra react-hook-form-oppsett
  - Typet `useForm`/`useFieldArray` med `FormValuesProps` og `Resolver<FormValuesProps>`
  - Strammere typing av feilhåndtering fra trygdeavgift-API (`unknown` med type guards)
  - Bruker riktig path-syntax for trigger-calls (`.` i stedet av `[]`)
  - Konverterte Vedtak-steg til TypeScript

  ### 📝 Endringer

  #### Flyttet filer
  - `src/sider/ftrl/saksbehandling/stegKomponenter/vurderingTrygdeavgift/` → `src/felleskomponenter/trygdeavgift/vurderingTrygdeavgift/`
  - Oppdatert alle imports i FTRL og EU/EØS steg-lister

  #### Nye filer
  - `src/felleskomponenter/stegvelger/StegState/tilStegstateMapping/trygdeavgiftStatus/index.ts`
  - `src/sider/eu_eøs/saksbehandling/stegMap/arbeid_tjenesteperson_eller_fly_vedtak.test.ts`

  #### Konvertert til TypeScript
  - `arbeid_tjenesteperson_eller_fly_vedtak.js` → `.ts`

  #### Oppdaterte filer
  - `Stegvelger.jsx` - Ny steg-rebuild logikk for trygdeavgiftStatus, ny store for trygdeavgiftStatus
  - `trygdeavgift.ts` - Ny status-håndtering, kriterier basert på `trygdeavgiftStatus`
  - `periode.ts` - Kriterier basert på om lovvalgsbestemmelse er satt (styrer visning av Trygdeavgift)
  - `vurderingTrygdeavgift.tsx` - `dataErLastet`-fix for doble API-kall, type-sikkerhet, forbedret form-håndtering
  - Diverse tester oppdatert for ny logikk

  ### 🧪 Testing
  - ✅ Alle eksisterende tester oppdatert og passerer
  - ✅ Nye tester for Vedtak-steget
  - ✅ Oppdaterte tester for Trygdeavgift og Periode
  - ✅ Manuell testing av steg-navigasjon
  - ✅ Verifisert at radio-knapp ikke lenger hopper tilbake
  - ✅ Verifisert kun ett API-kall ved interaksjon

  Hovedforskjellene:
  - ✅ Lagt til radio-knapp og doble API-kall problemer
  - ✅ Lagt til rotårsak (validering før data lastet)
  - ✅ Lagt til dataErLastet-fixen i løsningen
  - ✅ Lagt til flytting av filer til felleskomponenter
  - ✅ Fjernet "Forbedret form-håndtering" og "Optimalisert re-render-logikk" som ikke er relevante lenger
  - ✅ Forenklet type-sikkerhet-beskrivelsen

ragnarwestad committed:
7659 7661 Implementerte betinget visning av Trygdeavgift-steg

  Endret Periode- og Trygdeavgift-steg til å bruke dynamisk status-beregning,
  slik at Trygdeavgift-steget bare vises etter at Periode-steget er fullført.

  **Endringer i periode.ts:**
  - Endret kriterium exec() fra `() => true` til `() => !!this._propsLight.lovvalgsbestemmelse`
  - Fjernet hardkodet `this.status = FANE_STATUS.OK` i konstruktør
  - Status beregnes nå dynamisk av hentStatus() basert på harAvklaring:
    * UBEHANDLET når lovvalgsbestemmelse mangler
    * OK når lovvalgsbestemmelse er satt
  - Flyttet harAvklaring-beregning før kriterier-definisjon for bedre lesbarhet
  - Fjernet ubrukt import av FANE_STATUS

  **Endringer i trygdeavgift.ts:**
  - Fjernet hardkodet `this.status = FANE_STATUS.OK` i konstruktør
  - Status beregnes dynamisk av hentStatus() (alltid OK siden harAvklaring: true)
  - Fjernet ubrukt import av FANE_STATUS

  **Endringer i tester:**
  - periode.test.ts: Oppdatert 4 tester for å reflektere dynamisk status og betinget nesteSteg
  - trygdeavgift.test.ts: Oppdatert 1 test for å reflektere dynamisk status

  **Resultat:**
  - Trygdeavgift-steget vises nå bare når lovvalgsbestemmelse er valgt i Periode
  - Periode-steget viser UBEHANDLET status til lovvalgsbestemmelse er satt
  - Alle 1727 tester passerer
  - Build vellykket

ragnarwestad committed:
7659 7661 Fikset oppdaterRedux-flagg i trygdeavgift-status for å unngå state-reset

  Endret lagTrygdeavgiftStatus til å sette oppdaterRedux: false i stedet for true.
  Dette forhindrer at komponenten mister sin interne state når status oppdateres.

  Problem:
  - Når oppdaterRedux: true ble kalt publiserStegdata() som trigget Redux update
  - Redux update → full re-render av steg-systemet
  - VurderingTrygdeavgift-komponenten unmountet og remountet
  - All intern state (skjemaverdier) ble resatt
  - Brukeren mistet input umiddelbart etter å ha tastet/valgt noe

  Løsning:
  - Sett oppdaterRedux: false i lagTrygdeavgiftStatus
  - Status lagres kun i StegStore lokalt (Stegvelger state)
  - Ingen Redux update → ingen unødvendig re-render
  - Komponenten beholder sin interne state
  - Status oppdateres fortsatt dynamisk via beregnRelevantUI

  Trygdeavgift-status er UI-state, ikke forretningslogikk-state, og skal derfor
  ikke publiseres til Redux. Den brukes kun for å vise riktig fane-status i
  stegvalgeren.

ragnarwestad committed:
7659 7661 Optimaliserte trygdeavgift-status for å unngå unødvendige re-renders

  Implementerte optimalisering for å forhindre at trygdeavgift-komponenten
  resettes unødvendig ved status-oppdateringer.

  Endringer:
  - La til _sisteStatus-tracking i Trygdeavgift-klassen for å unngå
    duplikate oppdaterStegData-kall når status ikke har endret seg
  - Oppdaterte oppdaterStatus-handler til å kun kalle oppdaterStegData
    når isValid faktisk har endret seg fra forrige verdi
  - Utvidet test suite med test som verifiserer at oppdaterStegData
    ikke kalles ved uendret status

  Dette forhindrer unødvendige re-renders og sikrer at brukerens
  input-state i skjemaet ikke går tapt ved

ragnarwestad committed:
7659 7661 Fikset valideringsfeil som vises før data er lastet i trygdeavgift-steget

  Løste problemet der "Må fylles inn" feilmeldinger vises på skatteforhold-datoer
  når man navigerer tilbake til Trygdeavgift-steget.

  Problemet:
  - React Hook Form validerte umiddelbart ved component mount (mode: "onChange")
  - useEffect som kjører trigger() for re-validering ble utført FØR asynkron
    datalasting fra backend var ferdig
  - Dette resulterte i feilmeldinger på tomme felter selv om data eksisterte
    i databasen

  Løsning:
  - La til dataErLastet state-variabel for å spore når backend-data er lastet
  - Satt dataErLastet til false ved start av data-lasting (linje 173)
  - Satt dataErLastet til true etter at skjemaet er populert med data (linje 228)
  - Oppdatert useEffect (linje 314) til å kun kjøre trigger() når dataErLastet
    er true

  Dette sikrer at validering kun kjører etter at skjemaet er korrekt populert
  med data fra backend.

ragnarwestad committed:
7659 7661 Konverterte trygdeavgiftStatus til TypeScript og fjernet ubrukt kode

  Konverterte index.js til index.ts, la til TypeScript-typer, og fjernet
  ubrukte funksjoner for å følge prosjektets standard om å bruke TypeScript
  på all ny kode og unngå død kode.

  Endringer:
  - Renamed index.js til index.ts
  - La til import av StegData-type fra stegMotor/typer
  - La til type-annotasjon for lagTrygdeavgiftStatus: (isValid: boolean): StegData
  - Fjernet ubrukte funksjoner:
    - slettTrygdeavgiftStatus (ingen brukere)
    - konverterTrygdeavgiftStatusTilStegData (ingen brukere)
  - Fjernet ubrukte imports av konverterEnkelDataTilStegData og slettEnkelData

  Dette gir bedre type-sikkerhet, følger prosjektets kodestandarder for
  TypeScript-bruk, og holder kodebasen fri for død kode.

ragnarwestad committed:
7659 7661 Fjernet alle any-typer fra trygdeavgift-implementasjonen

  Forbedret type-sikkerhet ved å legge til trygdeavgiftStatus i PropsLight
  interface og fjerne alle any-casts i både produksjonskode og tester.

  Endringer:
  - La til trygdeavgiftStatus?: boolean i PropsLight interface (typer.ts:74)
  - Fjernet any-cast fra trygdeavgift.ts linje 41:
    Før: const trygdeavgiftStatus = (_propsLight as any).trygdeavgiftStatus;
    Nå:  const trygdeavgiftStatus = _propsLight.trygdeavgiftStatus;
  - Refaktorerte 4 tester til å bruke createMockPropsLight med type-safe
    trygdeavgiftStatus parameter i stedet for (propsLight as any).trygdeavgiftStatus

  Dette gir full type-sikkerhet i hele trygdeavgift-implementasjonen uten
  bruk av any-typer.

ragnarwestad committed:
7659 7661 Fikset dynamisk visning av Vedtak-steg og forbedret type-sikkerhet

  Hovedproblem:
  - Vedtak-steget dukket ikke opp umiddelbart når Trygdeavgift-steget ble godkjent
  - Det viste seg først ved navigering til et annet steg eller klikk på "Bekreft og fortsett"

  Rotårsak:
  - oppdaterStegData() i Stegvelger kalte kun oppdaterAktuelleSteg() når oppdaterRedux=true
  - Trygdeavgift brukte oppdaterRedux=false for å unngå unødvendige re-renders
  - Når trygdeavgiftStatus endret seg, ble ikke steg-kjeden rebuildet
  - Trygdeavgift-stegets kriterier ble derfor ikke re-evaluert før neste navigering

  Løsning:
  - La til else-branch i oppdaterStegData() som rebuilder steg-kjeden
  - Kun for type="trygdeavgiftStatus" for å unngå concurrency-problemer
  - Bruker debouncedOppdaterAktuelleSteg() (300ms) for å unngå multiple API-kall
  - Steg-kjeden rebuildes nå når trygdeavgiftStatus endres, uten database-feil

  Type-sikkerhet:
  - Fjernet any-typer fra react-hook-form-oppsett
  - Typet useForm/useFieldArray med FormValuesProps
  - Strammere typing av feilhåndtering fra trygdeavgift-API
  - Fikset inkonsistent håndtering av skalResetteFormverdier i hentOpprinneligTrygdeavgiftsgrunnlag()

  Endringer:
  - Forbedret håndtering av form-reset for å unngå å overskrive brukerens endringer
  - La til validering av formdata før API-kall for å unngå backend-feil
  - Bruker riktig path-syntax for trigger-calls (. i stedet for [])

ragnarwestad committed:
7659 7661 Fikset TypeScript intersection type for oppdaterStegData

  Erstattet ugyldig intersection type (&) med function overloads via interface.

  Problem:
  - Intersection type (&) for funksjoner krever at alle signaturer er kompatible samtidig
  - Dette fungerer ikke for function overloads der vi vil støtte to ulike signaturer

  Løsning:
  - Opprettet OppdaterStegDataHandler interface med function overloads
  - Støtter både (stegId, felt, verdi) og (stegId, data: StegData) signaturer
  - Dette er den korrekte måten å definere function overloads i TypeScript

CK-Hui commented:
Ser ut som det skjedde ganske mye i `src/sider/ftrl/saksbehandling/stegKomponenter/vurderingTrygdeavgift/vurderingTrygdeavgift.tsx` er refaktoreringen nødvendig for filen? Den rører ganske mange flyter så jeg lurer på om det er best å ikke gjøre det i denne PR.

Også må du huske å tagge denne PR med TOGGLE!

CK-Hui commented on src/felleskomponenter/stegvelger/StegState/tilStegstateMapping/trygdeavgiftStatus/index.ts:
```diff
@@ -0,0 +1,12 @@
+
+// Lagrer status uten å publisere til Redux (oppdaterRedux: false)
+// Dette forhindrer unødvendige re-renders som ville resette komponentens state
+export const lagTrygdeavgiftStatus = (isValid: boolean): StegData => ({
```
Et bedre navn blir kanskje `lagTrygdeavgiftBeregningStatus` må også da oppdateres for de andre.

ragnarwestad commented:
> Ser ut som det skjedde ganske mye i `src/sider/ftrl/saksbehandling/stegKomponenter/vurderingTrygdeavgift/vurderingTrygdeavgift.tsx` er refaktoreringen nødvendig for filen? Den rører ganske mange flyter så jeg lurer på om det er best å ikke gjøre det i denne PR.
> 
> Også må du huske å tagge denne PR med TOGGLE!

Har lagt til TOGGLE.

Forklaring følger, laget med hjelp fra "en god venn av meg" :-)
Det jeg er usikker på er om jeg har fulgt de vanlige mønstrene og "best practices" for steg i koden , så håper dere kan 
se på det og gi meg korrigeringer.

Endringene i vurderingTrygdeavgift.tsx kan deles inn i tre hovedkategorier:

  1. TypeScript type-sikkerhet (eliminering av any-typer)

  - useForm: Lagt til eksplisitt FormValuesProps type og Resolver<FormValuesProps>
  - useFieldArray: Fjernet as any og brukt eksplisitte typer FormValuesProps, "skatteforholdsperioder" og FormValuesProps, "inntektskilder"
  - mapFeilmelding: Komplett reskriving fra any parameter til unknown med proper type guards og type assertions
  - formValues iterasjon: Endret fra _periode: any til _periode: Skatteforhold og _periode: Inntektskilde
  - Import-rekkefølge: Flyttet useSelector til slutt for å følge konvensjon

  2. Fikset valideringsfeil som vises før data er lastet (src/sider/ftrl/saksbehandling/stegKomponenter/vurderingTrygdeavgift/vurderingTrygdeavgift.tsx:87-427)

  Dette var en kritisk bugfix:

  Problem: Validering ble trigget før data var ferdig lastet fra backend, noe som forårsaket feilmeldinger.

  Løsning:
  - Ny state-variabel dataErLastet for å tracke om initial data er hentet
  - håndterTrygdeavgiftsberegning og hentOpprinneligTrygdeavgiftsgrunnlag tar nå parameter skalResetteFormverdier for å unngå å overskrive brukerens endringer
  - useEffect på linje 427 kjører bare når dataErLastet === true

  3. Forbedret validering før API-kall (src/sider/ftrl/saksbehandling/stegKomponenter/vurderingTrygdeavgift/vurderingTrygdeavgift.tsx:269-350)

  Omfattende refaktorering av beregnTrygdeavgiftsperioder-logikken:

  Tidligere: Enkel if-else som sendte data til backend
  Nå: Grundig validering før API-kall med early returns:
  - Sjekker at skatteforholdsperioder eksisterer
  - Validerer at alle skatteforhold har påkrevde felt (fomDato, tomDato, skatteplikttype)
  - Sjekker at inntektskilder eksisterer når de er påkrevd
  - Validerer at alle inntektskilder har påkrevde felt
  - Forhindrer unødvendige API-kall med ufullstendige data

  Dette gjør koden mer robust og gir bedre brukerfeedback.

----

Begge bug'ene eksisterte i den opprinnelige JavaScript-koden og ble ikke introdusert av TypeScript-konverteringen. De ble bare oppdaget og fikset som en del av TypeScript-arbeidet.

  Bevis fra master-branchen:

  Bug 1: Validering kjører før data er lastet
  - Master-koden mangler dataErLastet state (linje 79 på din branch)
  - useEffect på linje 312-327 i master kjører alltid når aktivtSteg eller harEndretInnvilgetMedlemskapsperiode endres
  - Det betyr validering trigger umiddelbart, selv om data ennå ikke er hentet fra backend

  Bug 2: Ingen validering før API-kall
  - Master-koden (linje 246-268) sender data til beregnTrygdeavgiftsperioder uten å sjekke om feltene er komplette
  - Den bruker en enkel if-else som går rett til API-kallet
  - Din versjon legger til omfattende validering (linje 275-326) som forhindrer API-kall med ufullstendige data

  TypeScript-arbeidet hjalp med å finne bug'ene fordi:
  - Når du fjerner any-typer må du forstå datastrukturen bedre
  - Det tvinger deg til å se på alle steder hvor data flyter
  - Du oppdager edge cases som var skjult bak any-casting

Så TypeScript-konverteringen var faktisk nyttig - den avslørte eksisterende bugs som ikke var synlige i JavaScript-koden!

thanglyphan commented on src/felleskomponenter/stegvelger/Stegvelger.jsx:
```diff
@@ -157,11 +158,17 @@ class Stegvelger extends Component {
-      this.publiserStegdata();
-    }
+    this.setState(stegStores, () => {
+      // Callback kjøres etter setState er ferdig
```
Vi trenger ikke å ha kommentarer i koden når koden er beskrivende i seg selv, synes koden her er ryddig nok uten.

ragnarwestad replied:
Godt poeng, skal ta en runde på det

thanglyphan commented on src/sider/eu_eøs/saksbehandling/stegMap/arbeid_tjenesteperson_eller_fly_vedtak.ts:
```diff
@@ -2,13 +2,21 @@ import Steg from "../../../../felleskomponenter/stegvelger/stegMotor/steg";
 import MKV from "../../../../melosyskodeverk";
 
+interface VedtakRelevanteData {
+  [key: string]: unknown; // Index signature for base class compatibility
```
Hva er det denne gjør, og hvorfor trenger vi dette?

ragnarwestad replied:
Filen eksisterte allerede som .js - dette er en JS→TS konvertering

Filen arbeid_tjenesteperson_eller_fly_vedtak.ts er ikke ny funksjonalitet, men en konvertering fra JavaScript til TypeScript (rename fra .js til .ts).

  Hva gjør filen?

Dette er en steg-definisjon i stegvelgeren for EU/EØS-saksbehandling. Den definerer "Vedtak"-steget for behandling av arbeid som tjenesteperson eller flygende personell.

  Den håndterer:
  - Hvilket React-komponent som skal vises (velger mellom ny og legacy-versjon basert på feature toggle)
  - Hvilke data steget trenger (samleRelevanteData)
  - Hvilke handlers/callbacks steget kan bruke
  - Spesialhåndtering av lovvalgsbestemmelse 11.3a → 11.5

  Hva ble endret i TS-konverteringen?

  Type-annotasjoner lagt til:
  - PropsLight og number på constructor
  - VedtakRelevanteData interface for return-type av samleRelevanteData
  - string, unknown, boolean | undefined på handler-parametere
  - Non-null assertions (this.id!) der TypeScript trengte det

  Ingen funksjonell endring - kun type-sikkerhet.

thanglyphan commented on src/felleskomponenter/trygdeavgift/vurderingTrygdeavgift/vurderingTrygdeavgift.tsx:
```diff
@@ -80,6 +80,7 @@ export function VurderingTrygdeavgift({ bekreft, tilbake, aktivtSteg, oppdaterSt
     undefined,
   );
+  const [dataErLastet, setDataErLastet] = useState(false);
 
```
Denne fila ligger i FTRL mappa og brukes i EU_EØS flyt, kanskje flytte denne til felles mappe

thanglyphan commented on src/sider/ftrl/saksbehandling/stegKomponenter/vurderingTrygdeavgift/vurderingTrygdeavgift.tsx:
```diff
@@ -246,45 +269,125 @@ export function VurderingTrygdeavgift({ bekreft, tilbake, aktivtSteg, oppdaterSt
+    }
+
+    // Valider at vi har nødvendige data før vi sender til backend
+    if (!formVerdier.skatteforholdsperioder || formVerdier.skatteforholdsperioder.length === 0) {
```
Dette kan valideres i schema. Når valideringen i schema er false, ikke kall beregnTrygdeavgiftsperioder.

Samme med alt det under her som bare har en "return;", dette bør valideres i schema.

thanglyphan reviewed (changes requested):
Siden den rører FTRL, er dette toggle? 

Trenger endringer på denne, se kommentarer.

CK-Hui commented:
Tenker det heller er best å tilbakestille endringene introdusert for vurderingTrygdeavgift, vil heller spare refaktorering for en annen dag med tanke på at den komponenten rører så mye ting i melosys. Skjønte ikke helt forklaringen til AI heller, den prøver å fikse på noe som ikke er feil, virker for meg at den bare har funnet på en bug å fikse.

ragnarwestad committed:
7659 7661 Flyttet vurderingTrygdeavgift til felleskomponenter

  Hva ble gjort:

  1. Flyttet hele mappen fra src/sider/ftrl/saksbehandling/stegKomponenter/vurderingTrygdeavgift/ til src/felleskomponenter/trygdeavgift/vurderingTrygdeavgift/
    - 3 filer: .tsx, .less, .jsx (schema)
  2. Oppdatert imports i 4 FTRL steg-lister:
    - stegListeYrkesaktivFlyt.tsx
    - stegListePensjonistFlyt.tsx
    - stegListeManglendeInnbetalingFlyt.tsx
    - stegListePensjonistManglendeInnbetalingFlyt.tsx
  3. Oppdatert imports i EU/EØS:
    - stegMap/trygdeavgift.ts
    - stegMap/trygdeavgift.test.ts
  4. Oppdatert imports i felleskomponenter:
    - komponenter/inntektskilder.tsx (schema-import)
  5. Fikset alle interne imports i vurderingTrygdeavgift.tsx (fra 6 nivåer opp til 3 nivåer)
  6. Verifisert: TypeScript kompilerer uten feil ✅

ragnarwestad committed:
7659 7661 Forbedret navn på funksjon + fjernet kommentarer i Stegvelger

ragnarwestad committed:
7659 7661 Fjernet redundant validering fra vurderingTrygdeavgift

Problemet:
  - beregnTrygdeavgiftsperioder hadde 60+ linjer med imperative valideringer
  - Sjekkene dupliserte logikk som allerede finnes i vurderingTrygdeavgiftSchema
  - Dette brøt DRY-prinsippet og ga to kilder til sannhet for validering

  Løsning:
  - Fjernet alle imperative sjekker (skatteforholdsperioder/inntektskilder)
  - Schema validerer allerede alt via yup (min(1), required på alle felt)
  - debounceBeregnTrygdeavgiftsperioder sjekker isValid før kall
  - API kalles bare når formIsValid === true

ragnarwestad commented:
> Tenker det heller er best å tilbakestille endringene introdusert for vurderingTrygdeavgift, vil heller spare refaktorering for en annen dag med tanke på at den komponenten rører så mye ting i melosys. Skjønte ikke helt forklaringen til AI heller, den prøver å fikse på noe som ikke er feil, virker for meg at den bare har funnet på en bug å fikse.

Da må vi nok se litt nærmere på hva problemet er, den er nok ikke helt fri for bugs

ragnarwestad committed:
7659 7661 Fikset dobbel API-kall og radio-knapp som hoppet tilbake i Trygdeavgift-steget og lagt på typer for any's

  Problem:
  - Radio-knapp for "Skal det betales trygdeavgift" hoppet tilbake til "Ja" etter klikk på "Nei"
  - Dobbelt API-kall ved brukerinteraksjon
  - Validering kjørte før data var ferdig lastet fra backend

  Rotårsak:
  - useEffect som trigger validering kjørte umiddelbart ved mount
  - Form-feltene ble reset med data fra backend etter at bruker hadde gjort endringer
  - Dette førte til at brukerens valg ble overskrevet

  Løsning:
  - La til dataErLastet state for å tracke når initial data er hentet
  - Setter dataErLastet=false ved start av data-henting (useEffect)
  - Setter dataErLastet=true etter at form er populert med backend-data
  - Validering kjører kun når dataErLastet=true
  - La til skalResetteFormverdier parameter for å unngå å overskrive brukerens endringer

  Endringer:
  - Lagt til dataErLastet state-variabel
  - Oppdatert useEffect for data-henting med skalResetteFormverdier-logikk
  - Oppdatert validerings-useEffect med dataErLastet-sjekk
  - Oppdatert håndterTrygdeavgiftsberegning og hentOpprinneligTrygdeavgiftsgrunnlag med skalResetteFormverdier parameter

  Resultat:
  - Radio-knappen hopper ikke lenger tilbake
  - Kun ett API-kall ved interaksjon
  - Validering kjører kun når data er klar

7659 7661 Revertert til "Eirik Bergande* 13/11/2025, 14:21 7689 DAPI skatt og inntekt ikke tidligere år (#2926)"

CK-Hui committed:
Rydder bort endringer

CK-Hui committed:
rydding

CK-Hui committed:
Rydding

CK-Hui committed:
merge master
